### PR TITLE
PNA layers

### PR DESCRIPTION
**Pull request by tonydavis629**
## Description

I know this PR is long but it's mostly docstrings! This is the 2nd PR for Principal Neighborhood Aggregation layers. I plan to write the docs when the ModularTorchModel class is written.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
- [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
- [x] Run `mypy -p deepchem` and check no errors
- [x] Run `flake8 <modified file> --count` and check no errors
- [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/pna_gnn.py [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+                bond_embedding += self.bond_embedding_list[i](
+                    edge_attr[:, i].long() + 1)
+            else:
+                bond_embedding += self.bond_embedding_list[i](
+                    edge_attr[:, i].long())
+
+        return bond_embedding
+
+
+class PNALayer(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/pna_gnn.py:118]: Can you add to the docs in this PR?

File: deepchem/models/torch_models/pna_gnn.py [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+    batch_norm_momentum : float, optional, default=0.1
+        Momentum for the batch normalization layers.
+    avg_d : Dict[str, float], optional, default={"log": 1.0}
+        Dictionary containing the average degree of the graph.
+    posttrans_layers : int, optional, default=2
+        Number of post-transformation layers.
+    pretrans_layers : int, optional, default=1
+        Number of pre-transformation layers.
+
+    >>> import dgl
```
Comment (rbharath) [inline deepchem/models/torch_models/pna_gnn.py:157]: Should be in an example section

File: deepchem/models/torch_models/pna_gnn.py [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+    x = x / scales
+    x = torch.cat([x.sin(), x.cos()], dim=-1)
+    x = torch.cat((x, orig_x), dim=-1) if include_self else x
+    return x.squeeze()
+
+
+EPS = 1e-5
+
+
+def aggregate_mean(h, **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/pna_gnn.py:384]: Can be in utils and documented

**Review by rbharath — COMMENTED** [review]
Looks mostly good, let's break into 2 PRs, one for utils, and one for the layers themselves.

File: deepchem/models/torch_models/pna_gnn.py (hunk 3 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/pna_gnn.py]: done

File: deepchem/models/torch_models/pna_gnn.py [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+        elif self.edge_features and not self.pairwise_distances:
+            z2 = torch.cat(
+                [edges.src['feat'], edges.dst['feat'], edges.data['feat']],
+                dim=-1)
+        else:
+            z2 = torch.cat([edges.src['feat'], edges.dst['feat']], dim=-1)
+        return {"e": self.pretrans(z2)}
+
+
+def fourier_encode_dist(x, num_encodings=4, include_self=True):
```
[the remark this answers is not in the document]
Comment (tonydavis629) [reply deepchem/models/torch_models/pna_gnn.py]: done

File: deepchem/models/torch_models/pna_gnn.py (hunk 2 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/pna_gnn.py:220]: done

File: deepchem/models/torch_models/pna_gnn.py (hunk 1 above)
Comment (tonydavis629) [reply deepchem/models/torch_models/pna_gnn.py:179]: done

**Review by rbharath — COMMENTED** [review]
Will need to wait for rebase but not seeing anything jumping out on review

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**